src/on_chain/on_chain_surplus.py

```diff
@@ -0,0 +1,296 @@
+
+    def decode_single_hash(self, settlement_hash):
+        """
+        Need a better name for this function. Goes through all settlements fetched, decodes orders,
+        gets AWS Bucket response and makes call to quasimodo for required orders.
+        """
+        try:
+            encoded_transaction = self.web_3.eth.get_transaction(settlement_hash)
+            decoded_settlement = DecodedSettlement.new(
+                self.contract_instance, encoded_transaction.input
+            )
+        except ValueError as except_err:
+            self.logger.error(
+                "Unhandled exception, possibly can't decode: %s", str(except_err)
+            )
+            return None
+        (winning_orders, bucket_response) = self.get_order_data_by_hash(settlement_hash)
+        if bucket_response is None:
+            return bucket_response
+        # there can be multiple orders/trades in a single settlement
+        Bucket_Response = deepcopy(bucket_response)
+        for trade in decoded_settlement.trades:
+            del bucket_response
+            bucket_response = deepcopy(Bucket_Response)
+            order_type = str(
+                "{0:08b}".format(trade["flags"])
+            )  # convert flags value to binary to extract L.S.B (Least Sigificant Byte)
+            winning_surplus = self.extract_win_trade_data(
+                decoded_settlement, trade, order_type[-1]
+            )
+            order_id = winning_orders[decoded_settlement.trades.index(trade)]["id"]
+            for key, order in bucket_response["orders"].items():
+                if order["id"] == order_id:
+                    bucket_response["orders"] = {key: order}
+                    break
+            self.logger.info(
+                "Settlement Hash: %s\nFor order: %s\nWinning Surplus: %s",
+                settlement_hash,
+                order_id,
+                str(winning_surplus),
+            )
+
+            # convert back to JSON for sending to Quasimodo
+            # bucketResponseJson = json.dumps(bucket_response)
+            # # space here to post and receive instance JSON from Quasimodo
+            # instanceJson = requests.post(bucketResponseJson)
+            # assuming jsonObject is called instanceJson
+            # instance = instanceJson.json() to convert to dict
+            # 'instance' is what we use here as a python object in instance_file.py
+
+            if len(instance2["prices"]) > 0:
+                qmdo_surplus = self.extract_qmdo_trade_data(
+                    trade, order, instance2, order_type[-1]
+                )
+                diff_surplus = winning_surplus - qmdo_surplus
+                (percent_deviation, diff_in_eth) = self.get_conversions(
+                    diff_surplus,
+                    trade,
+                    order_type[-1],
+                    bucket_response["tokens"],
+                    order,
+                )
+                if percent_deviation < 0.1 and diff_in_eth < 0.002:
```

Comment:
Can we have this 0.1 and 0.002 as constants that are used throughout, so that if we need to change them, we would only look at one place and not all over the code in order to detect where this check happens.